src/compiler/compiler.rs

```diff
@@ -921,129 +944,174 @@ where
     T: CommandCreatorSync,
 {
     trace!("detect_compiler: {}", executable.display());
-
     // First, see if this looks like rustc.
 
-    let rustc_executable = if is_rustc_like(executable) {
-        Some(executable.to_path_buf())
+    let mut must_be_rust = false;
+    let mut rustc_executable = executable.to_path_buf();
+    if is_rustc_like(executable) {
+        must_be_rust = true;
     } else if env.iter().any(|(k, _)| k == OsStr::new("CARGO")) {
         // If not, detect the scenario where cargo is configured to wrap rustc with something other than sccache.
         // This happens when sccache is used as a `RUSTC_WRAPPER` and another tool is used as a
         // `RUSTC_WORKSPACE_WRAPPER`. In that case rustc will be the first argument rather than the command.
         //
         // The check for the `CARGO` env acts as a guardrail against false positives.
         // https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-reads
-        args.iter().next().and_then(|arg1| {
+        if let Some(arg1) = args.iter().next() {
             if is_rustc_like(arg1) {
-                Some(PathBuf::from(arg1))
-            } else {
-                None
+                must_be_rust = true;
+                rustc_executable = PathBuf::from(arg1)
             }
-        })
-    } else {
-        None
+        }
     };
 
-    let creator1 = creator.clone();
     let pool = pool.clone();
-    let cwd = cwd.to_owned();
-    if let Some(rustc_executable) = rustc_executable {
-        // Sanity check that it's really rustc.
-        let mut child = creator.clone().new_command_sync(executable);
-        // We're wrapping rustc if the executable doesn't match the detected rustc_executable. In this case the wrapper
-        // expects rustc as the first argument.
-        if rustc_executable != executable {
-            child.arg(&rustc_executable);
-        }
-
-        child.env_clear().envs(ref_env(env)).args(&["-vV"]);
-
-        let rustc_vv = run_input_output(child, None).await.map(|output| {
-            if let Ok(stdout) = String::from_utf8(output.stdout.clone()) {
-                if stdout.starts_with("rustc ") {
-                    return Ok(stdout);
-                }
+    if !must_be_rust && is_like_c_compiler(&rustc_executable) {
+        let executable = executable.to_owned();
+        let cc = detect_c_compiler(creator, executable, args, env.to_vec(), pool).await;
+        return cc.map(|c| (c, None));
+    }
+
+    // Even if it does not look like rustc like it might still be rustc driver
+    // so we do full check
+    match verified_rust_compiler(
```

Comment:
if I understand correctly, this function is performing an action

So, it should probably called
verify_rust_compiler